File: backend/app/schemas/combat.py
from pydantic import BaseModel
from typing import Optional
from app.models.combat import RangeBand


class FleetMemberCreate(BaseModel):
    ship_id: int
    count: int = 1


class FleetMemberRead(BaseModel):
    id: int
    ship_id: int
    count: int
    model_config = {"from_attributes": True}


class FleetCreate(BaseModel):
    name: str
    members: list[FleetMemberCreate] = []


class FleetRead(BaseModel):
    id: int
    name: str
    members: list[FleetMemberRead] = []
    model_config = {"from_attributes": True}


class BattleCreate(BaseModel):
    name: str
    fleet_a_id: int
    fleet_b_id: int


class BattleRead(BaseModel):
    id: int
    name: str
    fleet_a_id: int
    fleet_b_id: int
    current_round: int
    active: bool
    state: dict
    model_config = {"from_attributes": True}


class AttackAction(BaseModel):
    battle_id: int
    attacker_ship_id: int
    target_ship_id: int
    weapon_id: int
    gunner_skill: int = 0
    pilot_skill: int = 0
    dogfight_dm: int = 0
    evasive_target: bool = False   # target declared evasive action this round
    range_band: RangeBand


class ManeuverAction(BaseModel):
    battle_id: int
    ship_id: int
    thrust_spent: int
    current_range_band: RangeBand
    target_range_band: RangeBand

File: frontend/src/api/combat.ts
import { api } from "./client";
import type { Fleet, Battle, AttackResult, RangeBand } from "../types";

export const fetchFleets = () => api.get<Fleet[]>("/combat/fleets").then(r => r.data);
export const createFleet = (data: { name: string; members: { ship_id: number; count: number }[] }) =>
  api.post<Fleet>("/combat/fleets", data).then(r => r.data);

export const fetchBattles = () => api.get<Battle[]>("/combat/battles").then(r => r.data);
export const fetchBattle = (id: number) => api.get<Battle>(`/combat/battles/${id}`).then(r => r.data);
export const createBattle = (data: { name: string; fleet_a_id: number; fleet_b_id: number }) =>
  api.post<Battle>("/combat/battles", data).then(r => r.data);

export const resolveAttack = (data: {
  battle_id: number;
  attacker_ship_id: number;
  target_ship_id: number;
  weapon_id: number;
  gunner_skill: number;
  pilot_skill: number;
  range_band: RangeBand;
}) => api.post<AttackResult>("/combat/attack", data).then(r => r.data);
